man/man1/ra.1
.\" Copyright (c) 2000-2008 QoSient, LLC
.\" All rights reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\" 
.
.de TQ
.  br
.  ns
.  TP \\$1
..
.TH RA 1 "12 November 2007" "ra 3.0"
.SH NAME
\fBra\fP \- read \fBargus(8)\fP data.
.SH COPYRIGHT
Copyright (c) 2000-2008 QoSient. All rights reserved.
.SH SYNOPSIS
\fBra [raoptions] [- filter-expression]\fP
.SH DESCRIPTION
.IX  "ra command"  ""  "\fLra\fP \(em argus data"
.LP
.B Ra
reads
.BR argus(8)
data from either \fIstdin\fP, an \fIargus-file\fP, or from a
remote data source, which can either be an \fIargus-server\fP, 
or a netflow data server, filters the records it encounters
based on an optional \fIfilter-expression\fP and either prints
the contents of the
.BR argus(5)
records that it encounters to \fBstdout\fP or appends them into an
.B argus(5)
datafile.
.LP
.SH OPTIONS
.TP 4 4
.B \-A
Print aggregate statistics for the input stream on termination.
.TP 4 4
.B \-b
Dump the compiled transaction-matching code to standard output and stop.
This is useful for debugging filter expressions.
.TP 4 4
.B \-c <char>
Specify a delimiter character for output columns (default is ' ').
.TP 4 4
.B \-C <[host]:portnum>
Specify a source of Netflow data. The optional host is the local
interface address where Netflow Cisco records are going to be read.
If absent, then it is implied that the interface address is AF_ANY.
.TP 4 4
.B \-D <level>
Print debug information corresponding to \fB<level>\fP to stderr, if program
compiled to support debug printing.  As the level increases, so does the
amount of debug information
.B ra(1)
will print.  Values range from 1-8.
.TP 4 4
.B \-e <regex>
Match regular expression in flow user data fields.  Prepend the regex with
either "s:" or "d:" to limit the match to either the source or destination
user data fields.  Examples include:
.nf
   "^SSH-"           - Look for ssh connections on any port.
   "s:^GET"          - Look for HTTP GET requests in the source buffer.
   "d:^HTTP.*Unauth" - Find unauthorized http response.
.fi
.TP 4 4
.B \-E <file>
When using a filter expression at the end of the command, this option will
cause
.B ra(1)
to append the records that are rejected by the filter into
.B <file>
.TP 4 4
.B \-F <conffile>
Use \fB<conffile>\fP as a source of configuration information.  The format of
this file is identical to \fBrarc(5)\fP.  The data read from \fB<conffile>\fP
overrides any prior configuration information.
.TP 4 4
.B \-h
Print an explanation of all the arguments. 
.TP 4 4
.B \-n
Modify number to name converstion.  This flag supports 3 states, specified
by the modulus of the number of -n flags set. The first \fB-n\fP will suppress
address to hostname lookups.  \fB-nn\fP will suppress port number to service
conversion and \fB-nnn\fP will suppress translation of protocol numbers to
names.  \fB-nnnn\fP will return you to full conversion.   Because this
indicator can be set in the .rarc file, multiple \fB-n\fP flags can be used
to specify to a specific state of number to name conversion.
.TP 4 4
.B \-M <mode [mode ...]>
Provide addition mode operators.  These are generally specific to the
individual ra* program, or a specific function. Available modes for ra()
are:

.nf
   poll            - successfully attach to remote data source and then exit
   rmon            - modify data to support unidiretional RMON stat reporting
   saslmech="mech" - specify a mandatory SASL mech
   TZ="tzset"      - specify a tzset(3) time zone specification
   xml             - print output in xml format.
   label="regex"   - match flow label with regex(3) regular expression.
   dsrs=dsrlist    - process these dsrs
      Where a dsrlist has the format:
         [+/-]dsr[,[+/-]dsr]

         Supported dsrs are:
           trans  transport information, such as source id and seq number.
           flow   flow key data (proto, saddr, sport, dir, daddr, dport)
           time   time stamp fields (stime, ltime).
           metric basic ([s|d]bytes, [s|d]pkts, [s|d]rate, [s|d]load)
           agr    aggregation stats (trans, avgdur, mindur, maxdur, stdev).
           net    network objects (tcp, esp, rtp, icmp data).
           vlan   VLAN tag data
           mpls   MPLS label data
           jitter Jitter data ([s|d]jit, [s|d]intpkt)
           ipattr IP attributes ([s|d]ipid, [s|d]tos, [s|d]dsb, [s|d]ttl)
           psize  packet size information
           mac    MAC addresses (smac, dmac)
           icmp   ICMP specific data (icmpmap, inode)
           encaps Flow encapsulation type indications
           tadj   Time adjustment data
           cor    Multi-probe correlation data
           cocode Country Codes
           asn    Autonomous System Number data
           suser  src user captured data bytes (suser)
           duser  dst captured user data bytes (duser)

      Examples are:
         -M dsrs=time,flow,metric
         -M dsrs=-suser,-duser

   sql="select"    - use "select" as select clause in mysql call.

.fi
Illegal modes are not detectable by the standard library, and so unexpected
results in command line parsing may occur if care is not taken with use
of this option.
.TP 4 4
.B \-N <num>
.B \-N <start-end>
.B \-N <start+num>
Process the first \fB<num>\fP records, the inclusive range \fB<start - end >\fP
of records, or process <num + 1> records starting at start.  These
are the range of records that match the input filter if any filter is used.
.TP 4 4
.B \-p <digits>
Print \fB<digits>\fP number of units of precision for floating point values.
.TP 4 4
.B \-q
Run in quiet mode. Configure Ra to not print out the contents of records.
This can be used for a number of maintenance tasks, where you would be
interested in the outcome of a program, or its progress, say with the -D
option, without printing each input record.
.TP 4 4
.B \-r [- | <file file ...>]
Read data from \fB<files>\fP in the order presented on the
commandline. '\fB\-\fP' denotes stdin.  If you want to read a set of
files and then, when done, read stdin, use multiple occurences of
the \fI-r\fP option.  Ra can read \fBgzip(1)\fP, \fBbzip2(1)\fP and 
\fBcompress(1)\fP compressed data files.
.TP 4 4
.B \-R <dir dir ...>
Recursively decend the directory and process all the regular
files that are encountered.  The function does not decend to links, or
directories that begin with '.'.  The feature, like the -r command,
does not do any file type checking.
.TP 4 4
.B \-s <[-][[+[#]]field[:len] ...>
Specify the \fBfields\fP to print. Ra uses a default printing field list, 
by specifying a field you can replace this list completely, or you can
modify the existing default print list, using the optional '-' and '+[#]'
form of the command.  The available fields to print are:

.nf
   srcid, stime, ltime, sstime, dstime, dstime, dltime, 
   trans, seq, flgs, dur, avgdur, stddev, mindur, maxdur,
   saddr, daddr, proto, sport, dport, stos, dtos, sdsb, ddsb,
   sco, dco, sttl, dttl, sipid, dipid, smpls, dmpls, svlan, dvlan,
   svid, dvid, svpri, dvpri, [s|d]pkts, [s|d]bytes,
   [s||d]appbytes, [s|d]load, [s|d]loss, [s|d]ploss, [s|d]rate,
   smac, dmac, dir, [s|d]intpkt, [s|d]intpktact, [s|d]intpktidl,
   [s|d]jit, [s|d]jitact, [s|d]jitidl, state, suser, duser, 
   swin, dwin, trans, srng, erng, stcpb, dtcpb, tcprtt, inode, 
   offset, spktsz, dpktsz, smaxsz, dmaxsz, sminsz, dminsz

.fi

.PD 0
.RS
.TP 12
.B srcid

ChangeLog
Thu Aug 27 22:27:19 EDT 2009
Fixes for ralabel() to properly insert country codes.
Fixes for racluster() to aggregate correctly fow country codes.

Tue Aug 25 19:04:23 EDT 2009
Changes to thisid and argusid processing for CISCO conversion
Changes to rahisto() for better control of output (added -c option support).

Thu Aug 13 13:20:15 EDT 2009
Fix for racluster configuration to enable preservation when specified
in the configuration.  Getting ready to fix problems where records
headers are corrupt.

Sun Aug  9 19:44:21 EDT 2009
Fix for printing duration correctly when we have all the timestamps lite up.

Fri Jul 31 11:47:29 EDT 2009
Fix for 2.0 record conversion.  rtcp dsr inserted with len of 1 and
ArgusNtoH was corrupting other DSRs trying to little-endian the field.

Wed Jul 29 13:46:38 EDT 2009
Wow, "-S 1.2.3.4" didn't work, so fixed that.

Wed Jul 22 00:46:11 EDT 2009
Fix for trans processing in rahisto().
Modification of 'autoid' support in rasqlinsert().

Thu Jul  9 15:35:05 EDT 2009
Fix for rasplit() bug where the $srcid directive is ignored in some
situations. Not good.

Wed Jul  8 15:49:13 EDT 2009
Fix for TCP synAck flows.  Extend appbyte check in case argus wasn't 
configured to generate appbytes.
Fix for ratop ^L key.

Fri Jun 19 12:20:24 EDT 2009
Fix for ASN for inode IP address little endian converstion problem.

Mon Jun 15 09:12:25 EDT 2009
Major fix for processing IPv6 ICMP data.
Major fix for processing User data DSR.
Fix to not print field delimiter when  printing fields.

Thu May 21 13:11:59 EDT 2009
New fix for rasplit when spliting on anything bigger than hours.
New faster strategy for dealing with daylight savings time issues.
Updates for rabins for new spliting strategies.

Fri May 15 13:26:09 EDT 2009
Mods to fix RA_FIELD_WIDTH behavior.

Thu May  7 12:47:06 EDT 2009
Massive changes for Geolocation.
AS number reporting (change as from short to int).
Improved DNS reporting formats so field number is constant.
Fixed merging record problem.

Mon Apr 13 17:55:47 EDT 2009
Modified RA_FIELD_WIDTH behavior with using '-c' on the command line.
Added RA_FIELD_WIDTH to rarc.

Fixed problem with RaPrintIndex and rasqlinsert.
Modified rasqlinsert() minimum field width to be 256, so that addresses
will always be big enough.

Modified RA_TIME_FORMAT string parsing ala Dan Klien, to get non time
oriented characters and escaped sequences parsed correctly.

Mon Apr  6 11:37:36 EDT 2009
Fixed v2 -> v3 conversion.  ArgusNetworkStruct len was a few bytes too long.
Fixed mode parsing for rasplit().
Fix for rasplit() finding filename on split 'd','w','m','y'.  Algorithm
chose GMT not local timezone.

Fri Apr  3 12:53:15 EDT 2009
Fixes for rasqlinsert() to better deal with strftime() based table names.
Fixed ratree() to work in argus-3.0.

Mon Mar 30 16:34:24 EDT 2009
Added table date specifications for rasqlinsert().
Fix for memory problem when opening lots of files or directories.
Added RA_TIMEOUT_INTERVAL to rarc file.
Fixed short DSR parsing.

Fri Mar 27 09:41:05 EDT 2009
Fix for tcprtt propagation during merge when first record
is a ARGUS_TCP_INIT type DSR.

Fix for little-endian treatment of Authoritative response
in DNS printing routines of radump().

Wed Mar 25 16:09:53 EDT 2009
Mods to complete library to fix issues with -M rmon option.
Problem caused duplicate records generated by "-M rmon"
to be corrected before being aggregated.  Big problem,
basically just doubling metrics, but not flow ids.

Mon Mar 23 14:51:21 EDT 2009
Mod to ArgusClientInit in rabins() to deal with calling without
a -B option to specify the range.  Seems to work now.

Fri Mar 13 14:53:46 EDT 2009
Minor changes for raservices and rauserdata.
May have broken "finding" possible protocol types, need
to work on this this  weekend.

Thu Mar 12 11:40:14 EDT 2009
Removed -lz from aclocal.m4 an configure.in.  Will need
to add it back soon.

Wed Mar 11 20:30:17 EDT 2009
Fixed record searching when the read buffer boundary is

clients/ra.c
/*
 * Argus Software
 * Copyright (c) 2000-2009 QoSient, LLC
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* 
 * $Id: //depot/argus/clients/clients/ra.c#42 $
 * $DateTime: 2009/08/11 01:08:37 $
 * $Change: 1782 $
 */

/*
 *
 * ra  - Read Argus 
 *       This program read argus output streams, either through a socket,
 *       a piped stream, or in a file, filters and optionally writes the
 *       output to a file, its stdout or prints the binary records to
 *       stdout in ASCII.
 *
 * written by Carter Bullard
 * QoSient, LLC
 */

#if defined(CYGWIN)
#define USE_IPV6
#endif

#include <unistd.h>
#include <stdlib.h>

#include <compat.h>

#include <rabins.h>
#include <argus_util.h>
#include <argus_client.h>
#include <argus_main.h>
#include <argus_filter.h>

#include <signal.h>
#include <ctype.h>

extern int ArgusTotalMarRecords;
extern int ArgusTotalFarRecords;

extern struct ArgusParserStruct *ArgusParser;



void
ArgusClientInit (struct ArgusParserStruct *parser)
{
   struct ArgusModeStruct *mode = NULL;
   parser->RaWriteOut = 0;

   if (!(parser->RaInitialized)) {
      (void) signal (SIGHUP,  (void (*)(int)) RaParseComplete);
      (void) signal (SIGTERM, (void (*)(int)) RaParseComplete);
      (void) signal (SIGQUIT, (void (*)(int)) RaParseComplete);
      (void) signal (SIGINT,  (void (*)(int)) RaParseComplete);

      if ((mode = parser->ArgusModeList) != NULL) {
         while (mode) {
            if (!(strncasecmp (mode->mode, "poll", 4)))
               parser->RaPollMode++;

            if (!(strncasecmp (mode->mode, "rmon", 4)))
               parser->RaMonMode++;

            mode = mode->nxt;
         }
      }

      if (parser->Pctflag) {
         parser->ArgusPassNum = 2;
         if ((parser->ArgusAggregator = ArgusNewAggregator(parser, "srcid")) == NULL)
            ArgusLog (LOG_ERR, "ArgusClientInit: ArgusNewAggregator error");
      }
      parser->RaInitialized++;
   }
}

void RaArgusInputComplete (struct ArgusInput *input) { return; }

void
RaParseComplete (int sig)
{
   if (sig >= 0) {
      if (!ArgusParser->RaParseCompleting++) {

         if (ArgusParser->Aflag) {
#if defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__APPLE_CC__) || defined(__APPLE__) || defined(HAVE_SOLARIS)
            printf (" Totalrecords %-8lld  TotalMarRecords %-8lld  TotalFarRecords %-8lld TotalPkts %-8lld TotalBytes %-8lld\n",
                          ArgusParser->ArgusTotalRecords, ArgusParser->ArgusTotalMarRecords, ArgusParser->ArgusTotalFarRecords,
                          ArgusParser->ArgusTotalPkts, ArgusParser->ArgusTotalBytes);
#else
            printf (" Totalrecords %-8Ld  TotalManRecords %-8Ld  TotalFarRecords %-8Ld TotalPkts %-8Ld TotalBytes %-8Ld\n",
                          ArgusParser->ArgusTotalRecords, ArgusParser->ArgusTotalMarRecords, ArgusParser->ArgusTotalFarRecords,
                          ArgusParser->ArgusTotalPkts, ArgusParser->ArgusTotalBytes);
#endif
         }

         fflush(stdout);
         ArgusShutDown(sig);
      }

#ifdef ARGUSDEBUG
      ArgusDebug (2, "RaParseComplete(caught signal %d)\n", sig);
#endif
      switch (sig) {
         case SIGHUP:
         case SIGINT:
         case SIGTERM:
         case SIGQUIT: {
            struct ArgusWfileStruct *wfile = NULL;

            if (ArgusParser->ArgusWfileList != NULL) {
               struct ArgusListObjectStruct *lobj = NULL;
               int i, count = ArgusParser->ArgusWfileList->count;

               if ((lobj = ArgusParser->ArgusWfileList->start) != NULL) {
                  for (i = 0; i < count; i++) {
                     if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
                        if (wfile->fd != NULL) {
#ifdef ARGUSDEBUG
                           ArgusDebug (2, "RaParseComplete: closing %s\n", wfile->filename);
#endif
                           fflush (wfile->fd);
                           fclose (wfile->fd);
                           wfile->fd = NULL;
                        }
                     }
                     lobj = lobj->nxt;
                  }
               }
            }
            exit(0);
            break;
         }
      }
   }
}


void
ArgusClientTimeout ()
{
#ifdef ARGUSDEBUG
   ArgusDebug (2, "ArgusClientTimeout()\n");
#endif
}

void
parse_arg (int argc, char**argv)
{}

void
usage ()
{
   extern char version[];

   fprintf (stderr, "Ra Version %s\n", version);
   fprintf (stderr, "usage: %s \n", ArgusParser->ArgusProgramName);
   fprintf (stderr, "usage: %s [options] -S remoteServer  [- filter-expression]\n", ArgusParser->ArgusProgramName);
   fprintf (stderr, "usage: %s [options] -r argusDataFile [- filter-expression]\n\n", ArgusParser->ArgusProgramName);

   fprintf (stderr, "options: -A                    print record summaries on termination.\n");
   fprintf (stderr, "         -b                    dump packet-matching code.\n");
   fprintf (stderr, "         -c <char>             specify a delimiter <char> for output columns.\n");
   fprintf (stderr, "         -C <[host]:port>      specify Cisco Netflow source.\n");
#if defined (ARGUSDEBUG)
   fprintf (stderr, "         -D <level>            specify debug level\n");
#endif
   fprintf (stderr, "         -e <regex>            match regular expression in flow user data fields.\n");
   fprintf (stderr, "                               Prepend the regex with either \"s:\" or \"d:\" to limit the match\n");
   fprintf (stderr, "                               to either the source or destination user data fields.\n");
   fprintf (stderr, "         -E <file>             write records that are rejected by the filter into <file>\n");
   fprintf (stderr, "         -F <conffile>         read configuration from <conffile>.\n");
   fprintf (stderr, "         -h                    print help.\n");
   fprintf (stderr, "         -M <option>           specify a Mode of operation.\n");
   fprintf (stderr, "            rmon               convert bi-directional flow data to RMON in/out stats\n");
   fprintf (stderr, "            poll               attach to remote server to get MAR and then disconnect\n");
   fprintf (stderr, "            xml                pritn output in xml format\n");
   fprintf (stderr, "            TZ='timezone'      set TZ environment variable with timezone string\n");
   fprintf (stderr, "            saslmech='mech'    specify the sasl mechanism to use for this connection\n");
   fprintf (stderr, "            label='str'        specify label matching expression\n");
   fprintf (stderr, "            dsrs='strip str'   specify input dsrs (see rastrip.1)\n");
   fprintf (stderr, "            sql='str'          use str as \"WHERE\" clause in sql call.\n");
   fprintf (stderr, "            disa               Use US DISA diff-serve encodings\n");
   fprintf (stderr, "            hex                process user data using hex encoding\n");
   fprintf (stderr, "            ascii              process user data using ascii encoding\n");
   fprintf (stderr, "            encode32           process user data using encode32 encoding\n");
   fprintf (stderr, "            encode64           process user data using encode64 encoding\n");
   fprintf (stderr, "         -n                    don't convert numbers to names.\n");
   fprintf (stderr, "         -N <num>              process the first <num> records in the stream\n");
   fprintf (stderr, "            <start-end>        process this inclusive range of matching records.\n");
   fprintf (stderr, "            <start+num>        process this number of matching records, starting at start.\n");
   fprintf (stderr, "         -p <digits>           print fractional time with <digits> precision.\n");
   fprintf (stderr, "         -q                    quiet mode. don't print record outputs.\n");
   fprintf (stderr, "         -r <file>             read argus data <file>. '-' denotes stdin.\n");
   fprintf (stderr, "         -R <dir>              recursively process files in directory\n");
   fprintf (stderr, "         -s [-][+[#]]field[:w] specify fields to print.\n");
   fprintf (stderr, "                   fields:     srcid, stime, ltime, sstime, dstime, sltime, dltime,\n");
   fprintf (stderr, "                               trans, seq, flgs, dur, avgdur, stddev, mindur, maxdur,\n");
   fprintf (stderr, "                               saddr, daddr, proto, sport, dport, stos, dtos, sdsb, ddsb\n");
   fprintf (stderr, "                               sco, dco, sttl, dttl, sipid, dipid, smpls, dmpls, svlan, dvlan\n");
   fprintf (stderr, "                               svid, dvid, svpri, dvpri, [s|d]pkts, [s|d]bytes,\n");
   fprintf (stderr, "                               [s||d]appbytes, [s|d]load, [s|d]loss, [s|d]ploss, [s|d]rate,\n");
   fprintf (stderr, "                               smac, dmac, dir, [s|d]intpkt, [s|d]jit, state, suser, duser,\n");
   fprintf (stderr, "                               swin, dwin, trans, srng, erng, stcpb, dtcpb, tcprtt, inode,\n");
   fprintf (stderr, "                               offset, smaxsz, dmaxsz, sminsz, dminsz\n");
   fprintf (stderr, "         -S <host[:port]>      specify remote argus and optional port number\n");
   fprintf (stderr, "         -t <timerange>        specify <timerange> for reading records.\n");
   fprintf (stderr, "                   format:     timeSpecification[-timeSpecification]\n");
   fprintf (stderr, "                               timeSpecification: [[[yyyy/]mm/]dd.]hh[:mm[:ss]]\n");
   fprintf (stderr, "                                                    [yyyy/]mm/dd\n");
   fprintf (stderr, "                                                    -%%d{yMdhms}\n");
   fprintf (stderr, "         -T <secs>             attach to remote server for T seconds.\n");
   fprintf (stderr, "         -u                    print time in Unix time format.\n");
#ifdef ARGUS_SASL
   fprintf (stderr, "         -U <user/auth>        specify <user/auth> authentication information.\n");
#endif
   fprintf (stderr, "         -w <file>             write output to <file>. '-' denotes stdout.\n");
   fprintf (stderr, "         -X                    don't read default rarc file.\n");
   fprintf (stderr, "         -z                    print Argus TCP state changes.\n");
   fprintf (stderr, "         -Z <s|d|b>            print actual TCP flag values.\n");
   fprintf (stderr, "                               <'s'rc | 'd'st | 'b'oth>\n");
   exit(1);
}

void RaProcessThisRecord (struct ArgusParserStruct *, struct ArgusRecordStruct *);

void
RaProcessRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *argus)
{
   switch (argus->hdr.type & 0xF0) {
      case ARGUS_MAR:
         RaProcessManRecord (parser, argus);
         break;

      case ARGUS_EVENT:
         RaProcessEventRecord (parser, argus);
         break;

      case ARGUS_NETFLOW:
      case ARGUS_FAR: {
         struct ArgusMetricStruct *metric = (void *)argus->dsrs[ARGUS_METRIC_INDEX];
         if (metric != NULL) {
            parser->ArgusTotalPkts  += metric->src.pkts;
            parser->ArgusTotalPkts  += metric->dst.pkts;
            parser->ArgusTotalBytes += metric->src.bytes;
            parser->ArgusTotalBytes += metric->dst.bytes;
         }

         if (parser->RaMonMode) {
            struct ArgusRecordStruct *tns = ArgusCopyRecordStruct(argus);
            struct ArgusFlow *flow;

            if ((flow = (void *)argus->dsrs[ARGUS_FLOW_INDEX]) != NULL) {
               flow->hdr.subtype &= ~ARGUS_REVERSE;
               flow->hdr.argus_dsrvl8.qual &= ~ARGUS_DIRECTION;
            }

            RaProcessThisRecord(parser, argus);
            ArgusReverseRecord(tns);

            if ((flow = (void *)tns->dsrs[ARGUS_FLOW_INDEX]) != NULL) {
               flow->hdr.subtype &= ~ARGUS_REVERSE;
               flow->hdr.argus_dsrvl8.qual &= ~ARGUS_DIRECTION;
            }

            RaProcessThisRecord(parser, tns);
            ArgusDeleteRecordStruct(parser, tns);

         } else {
            RaProcessThisRecord(parser, argus);
         }
      }
   }
}

void
RaProcessThisRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *argus)
{
   static char buf[MAXSTRLEN];

   switch (parser->ArgusPassNum)  {
      case 2: {
         if (parser->Pctflag) {
            if (parser->ns == NULL) {
               parser->ns = ArgusCopyRecordStruct(argus);
            } else {
               ArgusMergeRecords (parser->ArgusAggregator, parser->ns, argus);
            }
         }
         break;
      }

      case 1: {
         if (parser->ArgusWfileList != NULL) {
            struct ArgusWfileStruct *wfile = NULL;
            struct ArgusListObjectStruct *lobj = NULL;
            int i, count = parser->ArgusWfileList->count;
      
            if ((lobj = parser->ArgusWfileList->start) != NULL) {
               for (i = 0; i < count; i++) {
                  if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
                     int retn = 1;
                     if (wfile->filterstr) {
                        struct nff_insn *wfcode = wfile->filter.bf_insns;
                        retn = ArgusFilterRecord (wfcode, argus);
                     }
      
                     if (retn != 0) {
                        if ((parser->exceptfile == NULL) || strcmp(wfile->filename, parser->exceptfile)) {
                           struct ArgusRecord *argusrec = NULL;
                           static char sbuf[0x10000];
                           if ((argusrec = ArgusGenerateRecord (argus, 0L, sbuf)) != NULL) {
      #ifdef _LITTLE_ENDIAN
                              ArgusHtoN(argusrec);
      #endif
                              ArgusWriteNewLogfile (parser, argus->input, wfile, argusrec);
                           }
                        }
                     }
                  }
      
                  lobj = lobj->nxt;
               }
            }
      
         } else {
            if (!parser->qflag) {
               if (parser->Lflag && !(parser->ArgusPrintXml)) {
                  if (parser->RaLabel == NULL)
                     parser->RaLabel = ArgusGenerateLabel(parser, argus);
       
                  if (!(parser->RaLabelCounter++ % parser->Lflag))
                     printf ("%s\n", parser->RaLabel);
       
                  if (parser->Lflag < 0)
                     parser->Lflag = 0;
               }
      
               bzero (buf, sizeof(buf));
               ArgusPrintRecord(parser, buf, argus, MAXSTRLEN);
      
               fprintf (stdout, "%s", buf);
      
               if (parser->eflag == ARGUS_HEXDUMP) {
                  int i;
                  for (i = 0; i < MAX_PRINT_ALG_TYPES; i++) {
                     if (parser->RaPrintAlgorithmList[i] != NULL) {
                        struct ArgusDataStruct *user = NULL;
                        if (parser->RaPrintAlgorithmList[i]->print == ArgusPrintSrcUserData) {
                           int slen = 0, len = parser->RaPrintAlgorithmList[i]->length;
                           if (len > 0) {
                              if ((user = (struct ArgusDataStruct *)argus->dsrs[ARGUS_SRCUSERDATA_INDEX]) != NULL) {
                                 if (user->hdr.type == ARGUS_DATA_DSR) {
                                    slen = (user->hdr.argus_dsrvl16.len - 2 ) * 4;
                                 } else
                                    slen = (user->hdr.argus_dsrvl8.len - 2 ) * 4;
                                    
                                 slen = (user->count < slen) ? user->count : slen;
                                 slen = (slen > len) ? len : slen;
                                 ArgusDump ((const u_char *) &user->array, slen, "      ");
                              }
                           }
                        }
                        if (parser->RaPrintAlgorithmList[i]->print == ArgusPrintDstUserData) {
                           int slen = 0, len = parser->RaPrintAlgorithmList[i]->length;
                           if (len > 0) {
                              if ((user = (struct ArgusDataStruct *)argus->dsrs[ARGUS_DSTUSERDATA_INDEX]) != NULL) {
                                 if (user->hdr.type == ARGUS_DATA_DSR) {
                                    slen = (user->hdr.argus_dsrvl16.len - 2 ) * 4;
                                 } else
                                    slen = (user->hdr.argus_dsrvl8.len - 2 ) * 4;
      
                                 slen = (user->count < slen) ? user->count : slen;
                                 slen = (slen > len) ? len : slen;
                                 ArgusDump ((const u_char *) &user->array, slen, "      ");
                              }
                           }
                        }
                     } else
                        break;
                  }
               }
      
               fprintf (stdout, "\n");
               fflush (stdout);
            }
         }
      }
   }
}

void
RaProcessManRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *argus)
{
   static char buf[MAXSTRLEN];

   if (parser->ArgusWfileList != NULL) {
      struct ArgusWfileStruct *wfile = NULL;
      struct ArgusListObjectStruct *lobj = NULL;
      int i, count = parser->ArgusWfileList->count;

      if ((lobj = parser->ArgusWfileList->start) != NULL) {
         for (i = 0; i < count; i++) {
            if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
               int retn = 1;
               if (wfile->filterstr) {
                  struct nff_insn *wfcode = wfile->filter.bf_insns;
                  retn = ArgusFilterRecord (wfcode, argus);
               }

               if (retn != 0) {
                  if ((parser->exceptfile == NULL) || strcmp(wfile->filename, parser->exceptfile)) {
                     struct ArgusRecord *argusrec = NULL;
                     static char sbuf[0x10000];
                     if ((argusrec = ArgusGenerateRecord (argus, 0L, sbuf)) != NULL) {
#ifdef _LITTLE_ENDIAN
                        ArgusHtoN(argusrec);
#endif
                        ArgusWriteNewLogfile (parser, argus->input, wfile, argusrec);
                     }
                  }
               }
            }

            lobj = lobj->nxt;
         }
      }

   } else {

      if ((parser->ArgusPrintMan) && (!parser->qflag)) {
         if (parser->Lflag && !(parser->ArgusPrintXml)) {
            if (parser->RaLabel == NULL)
               parser->RaLabel = ArgusGenerateLabel(parser, argus);
 
            if (!(parser->RaLabelCounter++ % parser->Lflag))
               printf ("%s\n", parser->RaLabel);
 
            if (parser->Lflag < 0)
               parser->Lflag = 0;
         }

         bzero (buf, sizeof(buf));
         if (argus->dsrs[0] != NULL) {
            ArgusPrintRecord(parser, buf, argus, MAXSTRLEN);
            fprintf (stdout, "%s\n", buf);
         }
         fflush (stdout);
      }
   }

#ifdef ARGUSDEBUG
   {
      struct ArgusRecord *rec = (struct ArgusRecord *)argus->dsrs[0];
      if (rec != NULL) {
         struct ArgusMarStruct *mar = &rec->ar_un.mar;
         ArgusDebug (6, "RaProcessManRecord (0x%x, 0x%x) mar parsed 0x%x", parser, argus, mar); 
      }
   }
#endif
}


void
RaProcessEventRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *argus)
{
   static char buf[MAXSTRLEN];

   if (parser->ArgusWfileList != NULL) {
      struct ArgusWfileStruct *wfile = NULL;
      struct ArgusListObjectStruct *lobj = NULL;
      int i, count = parser->ArgusWfileList->count;

      if ((lobj = parser->ArgusWfileList->start) != NULL) {
         for (i = 0; i < count; i++) {
            if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
               int retn = 1;
               if (wfile->filterstr) {
                  struct nff_insn *wfcode = wfile->filter.bf_insns;
                  retn = ArgusFilterRecord (wfcode, argus);
               }

               if (retn != 0) {
                  if ((parser->exceptfile == NULL) || strcmp(wfile->filename, parser->exceptfile)) {
                     struct ArgusRecord *argusrec = NULL;
                     static char sbuf[0x10000];
                     if ((argusrec = ArgusGenerateRecord (argus, 0L, sbuf)) != NULL) {
#ifdef _LITTLE_ENDIAN
                        ArgusHtoN(argusrec);
#endif
                        ArgusWriteNewLogfile (parser, argus->input, wfile, argusrec);
                     }
                  }
               }
            }

            lobj = lobj->nxt;
         }
      }

   } else {

      if ((parser->ArgusPrintEvent) && (!parser->qflag)) {
         if (parser->Lflag && !(parser->ArgusPrintXml)) {
            if (parser->RaLabel == NULL)
               parser->RaLabel = ArgusGenerateLabel(parser, argus);
 
            if (!(parser->RaLabelCounter++ % parser->Lflag))
               printf ("%s\n", parser->RaLabel);
 
            if (parser->Lflag < 0)
               parser->Lflag = 0;
         }

         bzero (buf, sizeof(buf));
         ArgusPrintRecord(parser, buf, argus, MAXSTRLEN);

         fprintf (stdout, "%s\n", buf);
         fflush (stdout);
      }
   }

#ifdef ARGUSDEBUG
   {
      struct ArgusRecord *rec = (struct ArgusRecord *)argus->dsrs[0];

      if (rec != NULL) {
         struct ArgusEventStruct *event = &rec->ar_un.event;
         ArgusDebug (6, "RaProcessEventRecord (0x%x, 0x%x) event parsed 0x%x", parser, argus, event); 
      }
   }
#endif
}

int RaSendArgusRecord(struct ArgusRecordStruct *argus) {return 0;}

void ArgusWindowClose(void);

void ArgusWindowClose(void) { 
#ifdef ARGUSDEBUG
   ArgusDebug (6, "ArgusWindowClose () returning\n"); 
#endif
}
